File: mobile/src/api/safety.ts
import { apiFetch } from '@/src/api/client';

export type PretaskChecklist = {
  supervisor_walkthrough: boolean;
  coordination_other_crafts: boolean;
  equipment_check: boolean;
  training_complete: boolean;
  sufficient_personnel: boolean;
};

export type PretaskTask = {
  jha_complete: boolean;
  task: string;
  hazards: string;
  steps: string;
};

export type PretaskAttendee = {
  print_name: string;
  signature: string;
};

export type DailyPretask = {
  id: string;
  project_id: string;
  project_name: string;
  project_number: string | null;
  work_date: string;
  crew_lead_user_id: string;
  crew_lead_name: string;
  daily_report_id: string | null;
  client_id: string | null;
  company_name: string;
  area_of_work: string;
  status: 'draft' | 'submitted';
  checklist: PretaskChecklist;
  tasks: PretaskTask[];
  near_miss: boolean;
  near_miss_notes: string;
  required_permits: string;
  items_concerns: string;
  quality_previous_day: string;
  present_items_concerns: string;
  attendees: PretaskAttendee[];
  supervisor_name: string;
  supervisor_signature: string;
  submitted_at: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type DailyPretaskListItem = {
  id: string;
  project_id: string;
  project_name: string;
  project_number: string | null;
  work_date: string;
  area_of_work: string;
  status: 'draft' | 'submitted';
  crew_lead_name: string;
  task_count: number;
  attendee_count: number;
  near_miss: boolean;
  submitted_at: string | null;
  updated_at: string | null;
};

type PretaskResponse = { item: DailyPretask; entity: string; created?: boolean };
type PretaskListResponse = { items: DailyPretaskListItem[]; total: number; entity: string };

export type PretaskWrite = Partial<
  Pick<
    DailyPretask,
    | 'company_name'
    | 'area_of_work'
    | 'checklist'
    | 'tasks'
    | 'near_miss'
    | 'near_miss_notes'
    | 'required_permits'
    | 'items_concerns'
    | 'quality_previous_day'
    | 'present_items_concerns'
    | 'attendees'
    | 'supervisor_name'
    | 'supervisor_signature'
    | 'daily_report_id'
    | 'client_id'
  >
> & { work_date?: string; status?: 'draft' | 'submitted' };

export async function getOrCreateDailyPretask(
  projectId: string,
  workDate: string,
  clientId?: string,
): Promise<DailyPretask> {
  const qs = new URLSearchParams({ date: workDate });
  if (clientId) qs.set('client_id', clientId);
  const data = await apiFetch<PretaskResponse>(
    `/api/v1/projects/${projectId}/daily-pretasks?${qs.toString()}`,
  );
  return data.item;
}

export async function saveDailyPretask(id: string, body: PretaskWrite): Promise<DailyPretask> {
  const data = await apiFetch<PretaskResponse>(`/api/v1/daily-pretasks/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return data.item;
}

export async function submitDailyPretask(id: string, body: PretaskWrite = {}): Promise<DailyPretask> {
  const data = await apiFetch<PretaskResponse>(`/api/v1/daily-pretasks/${id}/submit`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  return data.item;
}

export async function listDailyPretasks(params?: {
  projectId?: string;
  date?: string;
}): Promise<DailyPretaskListItem[]> {
  const qs = new URLSearchParams();
  if (params?.projectId) qs.set('project_id', params.projectId);
  if (params?.date) qs.set('date', params.date);
  const suffix = qs.toString() ? `?${qs.toString()}` : '';
  const data = await apiFetch<PretaskListResponse>(`/api/v1/safety/pretasks${suffix}`);
  return data.items ?? [];
}

File: backend/app/models/safety.py
"""Project-scoped safety records (Plan 7 daily pretask)."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any, Optional

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..extensions import db
from .base import TimestampMixin, UUIDPKMixin

PRETASK_STATUSES = ("draft", "submitted")

PRETASK_CHECKLIST_KEYS = (
    "supervisor_walkthrough",
    "coordination_other_crafts",
    "equipment_check",
    "training_complete",
    "sufficient_personnel",
)

DEFAULT_COMPANY_NAME = "DOCON, INC"


def empty_pretask_checklist() -> dict[str, bool]:
    return {key: False for key in PRETASK_CHECKLIST_KEYS}


def empty_pretask_task() -> dict[str, Any]:
    return {"jha_complete": False, "task": "", "hazards": "", "steps": ""}


def empty_pretask_attendee() -> dict[str, str]:
    return {"print_name": "", "signature": ""}


def default_pretask_tasks() -> list[dict[str, Any]]:
    return [empty_pretask_task() for _ in range(4)]


def default_pretask_attendees() -> list[dict[str, str]]:
    return [empty_pretask_attendee() for _ in range(4)]


class DailyPretask(UUIDPKMixin, TimestampMixin, db.Model):
    """One daily pre-task safety plan per project, workday, and crew lead."""

    __tablename__ = "daily_pretasks"
    __table_args__ = (
        UniqueConstraint(
            "project_id",
            "work_date",
            "crew_lead_user_id",
            name="uq_daily_pretasks_project_date_lead",
        ),
        UniqueConstraint("client_id", name="uq_daily_pretasks_client_id"),
    )

    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    work_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    crew_lead_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    daily_report_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("daily_reports.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    client_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    company_name: Mapped[str] = mapped_column(String(255), nullable=False, default=DEFAULT_COMPANY_NAME)
    area_of_work: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="draft")
    checklist: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False, default=dict)
    tasks: Mapped[list[Any]] = mapped_column(JSONB, nullable=False, default=list)
    near_miss: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    near_miss_notes: Mapped[str] = mapped_column(Text, nullable=False, default="")
    required_permits: Mapped[str] = mapped_column(Text, nullable=False, default="")
    items_concerns: Mapped[str] = mapped_column(Text, nullable=False, default="")
    quality_previous_day: Mapped[str] = mapped_column(Text, nullable=False, default="")
    present_items_concerns: Mapped[str] = mapped_column(Text, nullable=False, default="")
    attendees: Mapped[list[Any]] = mapped_column(JSONB, nullable=False, default=list)
    supervisor_name: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    supervisor_signature: Mapped[str] = mapped_column(Text, nullable=False, default="")
    submitted_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    created_by_user_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    project = relationship("Project", foreign_keys=[project_id])
    crew_lead = relationship("User", foreign_keys=[crew_lead_user_id])
    created_by = relationship("User", foreign_keys=[created_by_user_id])
